src/benchmark_environments/classic_control.py

```diff
@@ -0,0 +1,37 @@
+"""Adaptation of classic Gym environments for IRL."""
+
+import gym.wrappers
+
+from benchmark_environments import util
+
+register = util.curried_gym_register_as_decorator(__name__)
+
+
+@register("CartPole-v0")
+def cart_pole():
+    """Fixed-length variant of CartPole-v1.
+
+    If the agent fails (IE pole falls over or CartPole exits the screen),
+    all rewards become 0 for the duration of the episode.
```

Comment:
This does mean the reward is non-Markovian which is a little strange (although plausibly fairly common in practice). Possible alternatives:
  1. Just let the reward be positive if it manages to get the pole upright again (it probably won't!)
  2. Use the `AbsorbAfterDoneWrapper`.

@qxcv do you have thoughts on this?

Reply:
I doubt it will make much difference to RL. That said, I'd prefer alternative (1) or (2) because those align with the original motivations of the benchmark reimp project (e.g. removing hidden state from reward functions, etc.).

Reply:
OK. In that case I think I'd lean towards (1) since this seems more physically realistic, but ok with (2) as well if it's much easier to implement.

Reply:
(1) seems fine to me. I have a clarifying question about what counts as upright which I'm asking on Slack

Reply:
It looks like the cleanest way to implement (1) is to subclass Gym's CartPoleEnv and override `step()`.

Reply:
I think rotating 360 degrees so it ends up vertically upwards again should still count as upright. This is analogous to a robotic "swing-up" task.

I can see a case for disallowing this, but I think we'd then need to augment the state space to indicate it if it ever fell over, otherwise the reward would be non-Markovian.

Hope you're feeling better!

Reply:
Two weird things I wanted to bring up:
(a) Under the original implementation (with termination condition removed), the meaning of the CW and CCW actions becomes reversed when the pole is under the cart. Is this worth fixing? (Don't expect RL algorithms to be hurt by this, but could be confusing. Fixing probably just means adding another if statement.)

(b) I'd expect that if I removed all force except gravity that the pole would swing back and forth in an arc, [but what actually happens](https://www.youtube.com/watch?v=MluW34l_g5I&feature=youtu.be) is that the pole completes the entire circle, and in fact starts accelerating with each completion. Even if I force initial theta velocity to be 0 and reduce tau by 10x ("seconds between state updates"). Probably due to the eulers method approximation. I'm not expecting this to affect RL.

Reply:
> I can see a case for disallowing this, but I think we'd then need to augment the state space to indicate it if it ever fell over, otherwise the reward would be non-Markovian.

trivia: Actually, in the original implementation but with terminal state removed, we are in a distinct state when we swing 360 degrees around. The pole theta appears as something like 2*pi rather than 0 when it is at the top because the angle is never normalized.

Reply:
> Two weird things I wanted to bring up:
> (a) Under the original implementation (with termination condition removed), the meaning of the CW and CCW actions becomes reversed when the pole is under the cart. Is this worth fixing? (Don't expect RL algorithms to be hurt by this, but could be confusing. Fixing probably just means adding another if statement.)

As in the CW and CCW actions exert torque in different directions depending on the y-coordinate? I'd be inclined to fix this so it's more physically realistic. Agree it probably won't affect RL (optimal policy is to never fall over, after all) though.

> (b) I'd expect that if I removed all force except gravity that the pole would swing back and forth in an arc, [but what actually happens](https://www.youtube.com/watch?v=MluW34l_g5I&feature=youtu.be) is that the pole completes the entire circle, and in fact starts accelerating with each completion. Even if I force initial theta velocity to be 0 and reduce tau by 10x ("seconds between state updates"). Probably due to the eulers method approximation. I'm not expecting this to affect RL.

Are you sure you haven't just invented a perpetual motion machine? ;) Yeah this doesn't seem worth fixing.

Reply:
> (a) Under the original implementation (with termination condition removed), the meaning of the CW and CCW actions becomes reversed when the pole is under the cart. Is this worth fixing? (Don't expect RL algorithms to be hurt by this, but could be confusing. Fixing probably just means adding another if statement.)

This was my misunderstanding -- there's aren't CW/CCW actions on the pole. Rather, there are left/right actions on the cart.